source/common/upstream/ring_hash_lb.cc

```diff
@@ -60,34 +62,115 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
-  // configuration in the future.
   // NOTE: Currently we keep a ring for healthy hosts and unhealthy hosts, and this is done per
   //       thread. This is the simplest implementation, but it's expensive from a memory
   //       standpoint and duplicates the regeneration computation. In the future we might want
   //       to generate the rings centrally and then just RCU them out to each thread. This is
   //       sufficient for getting started.
+
   const uint64_t min_ring_size =
-      config ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, 1024) : 1024;
+      config ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size,
+                                               DEFAULT_MIN_RING_SIZE)
+             : DEFAULT_MIN_RING_SIZE;
+  const uint64_t max_ring_size =
+      config ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), maximum_ring_size,
+                                               DEFAULT_MAX_RING_SIZE)
+             : DEFAULT_MAX_RING_SIZE;
+  const absl::optional<uint64_t> target_hashes_per_host =
+      config && config->has_target_hashes_per_host()
+          ? absl::optional<uint64_t>(config->target_hashes_per_host().value())
+          : absl::nullopt;
+
+  // Sanity-check ring size bounds.
+  if (min_ring_size > max_ring_size) {
+    ENVOY_LOG(error, "ring hash: minimum_ring_size ({}) > maximum_ring_size ({})", min_ring_size,
+              max_ring_size);
+    return;
+  }
 
-  uint64_t hashes_per_host = 1;
-  if (hosts.size() < min_ring_size) {
-    hashes_per_host = min_ring_size / hosts.size();
-    if ((min_ring_size % hosts.size()) != 0) {
-      hashes_per_host++;
+  // Compute the "effective" weight of each host as the product of its own weight and the locality
+  // weight, if given. Sum these effective weights and find their greatest common denominator for
+  // normalization.
+  // NOTE: GCD is associative, same as std::max() for example. To evaluate GCD(a, b, c, ...), we
+  //       compute GCD(GCD(GCD(GCD(0, a), b), c), ...).
+  uint64_t weighted_sum = 0;
+  uint32_t gcd = 0;
+  std::unordered_map<HostConstSharedPtr, uint32_t> effective_weights;
+  for (uint64_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      auto host_weight = host->weight();
+      ASSERT(host_weight != 0);
+      // NOTE: Locality weight may be zero. In this case, all hosts in the locality are assigned
+      //       no load. When we build the ring below, such hosts will have no entries in the ring.
+      // TODO: When we move to C++17, change this to `locality_weights[i]` (i.e. use
+      //       std::shared_ptr::operator[]) rather than dereferencing locality_weights explicitly.
+      auto locality_weight = locality_weights == nullptr ? 1 : (*locality_weights)[i];
+      auto effective_weight = host_weight * locality_weight;
+      weighted_sum += effective_weight;
+      gcd = Envoy::gcd(gcd, effective_weight);
```

Comment:
Nit: do you need the `Envoy::` prefix?

Reply:
I do, to disambiguate between the function and the local variable name that shadows it. It'd be the same with `std::gcd()`.